server/src/main/java/org/opensearch/tasks/TaskManager.java

```diff
@@ -116,10 +129,23 @@ public class TaskManager implements ClusterStateApplier {
     private final Map<TcpChannel, ChannelPendingTaskTracker> channelPendingTaskTrackers = ConcurrentCollections.newConcurrentMap();
     private final SetOnce<TaskCancellationService> cancellationService = new SetOnce<>();
 
-    public TaskManager(Settings settings, ThreadPool threadPool, Set<String> taskHeaders) {
+    private boolean taskResourceConsumersEnabled;
+    private final List<Consumer<Task>> taskResourceConsumer;
+
+    public TaskManager(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool, Set<String> taskHeaders) {
         this.threadPool = threadPool;
         this.taskHeaders = new ArrayList<>(taskHeaders);
         this.maxHeaderSize = SETTING_HTTP_MAX_HEADER_SIZE.get(settings);
+        this.taskResourceConsumersEnabled = TASK_RESOURCE_CONSUMERS_ENABLED.get(settings);
+        this.taskResourceConsumer = new ArrayList<Consumer<Task>>() {
+            {
+                add(new TopNSearchTasksLogger(settings));
+            }
+        };
+
+        if (clusterSettings != null) {
+            clusterSettings.addSettingsUpdateConsumer(TASK_RESOURCE_CONSUMERS_ENABLED, this::setTaskResourceConsumersEnabled);
+        }
```

Comment:
What are the cases clusterSettings can be null. The defensive check is scary since it gives a sense that the control knob might not be absolute

Reply:
TaskManager object is created from TransportService class which checks for clusterSettings nullability before adding the update consumer and hence this check is also added here. I think the intent behind adding this check is for tests, a lot of tests rely on TransportService but send empty ClusterSettings. 

```
        if (clusterSettings != null) {
            clusterSettings.addSettingsUpdateConsumer(TransportSettings.TRACE_LOG_INCLUDE_SETTING, this::setTracerLogInclude);
            clusterSettings.addSettingsUpdateConsumer(TransportSettings.TRACE_LOG_EXCLUDE_SETTING, this::setTracerLogExclude);
            if (remoteClusterClient) {
                remoteClusterService.listenForUpdates(clusterSettings);
            }
            clusterSettings.addSettingsUpdateConsumer(TransportSettings.SLOW_OPERATION_THRESHOLD_SETTING, transport::setSlowLogThreshold);
        }
```
https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/TransportService.java#L219